### PR TITLE
out_forward plugin can set targets dynamically via service discovery plugin.

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 
Fixes https://github.com/fluent/fluentd/issues/756 ?

**What this PR does / why we need it**: 

I Added service_discovery plugin which sets config of servers in out_forwardy dynamically.
In this PR, I supported `sd_file` plugins and `sd_static` plugin.
sd_file checks the specified file periodically and update its targets.
sd_static is same as `<server>` configuration. So the following configurations are completely the same.

```
<match test>
  @type forward

  <server>
    host 127.0.0.1
    port 24224
  </server>
<match>
```

```
<match test>
  @type forward

  <service_discovery>
    @type static
    host 127.0.0.1
    port 24224
  </server>
<match>
```

**Docs Changes**:

Needed for service discovery plugin. 

**Release Note**: 

out_forward plugin can set targets dynamically via service discovery plugin.

### CODE REVIEW

File: lib/fluent/plugin/sd_file.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+#
+
+require 'cool.io'
+
+require 'fluent/plugin/service_discovery'
+require 'fluent/plugin/service_discovery/discovery_message'
+
+module Fluent
+  module Plugin
+    class SdFile < ServiceDiscovery
```
Comment (repeatedly) [inline lib/fluent/plugin/sd_file.rb]: Use `FileServiceDiscovery` better

File: lib/fluent/plugin/sd_file.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            s['name'],
+            s.fetch('weight', DEFAUT_WEIGHT),
+            s['standby'],
+            s['username'],
+            s['password'],
+            s['shared_key'],
+          )
+        end
+      rescue KeyError => e
+        raise Fluent::ConfigError, e
```
Comment (repeatedly) [inline lib/fluent/plugin/sd_file.rb]: Maybe, `KeyError`'s error message is not user friendly.
Adding more information is better.

File: lib/fluent/plugin/sd_static.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'fluent/plugin/service_discovery'
+
+module Fluent
+  module Plugin
+    class SdStatic < ServiceDiscovery
```
Comment (repeatedly) [inline lib/fluent/plugin/sd_static.rb]: `StaticServiceDiscovery` is better

File: lib/fluent/plugin/sd_static.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      config_param :password, :string, default: '', secret: true
+      desc 'Marks a node as the standby node for an Active-Standby model between Fluentd nodes.'
+      config_param :standby, :bool, default: false
+      desc 'The load balancing weight.'
+      config_param :weight, :integer, default: 60
+
+      def configure(conf)
+        super
+
+        @services << ServiceDiscovery::Service.new(:static, @host, @port, @name, @weight, @standby, @username, @password, @shared_key)
```
Comment (repeatedly) [inline lib/fluent/plugin/sd_static.rb]: Does this plugin support only one server unlike out_forward?

File: lib/fluent/plugin/service_discovery/round_robin_balancer.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+          @mutex = Mutex.new
+        end
+
+        def rebalance(services)
+          @mutex.synchronize do
+            @services = services
+          end
+        end
+
+        def select_node
```
Comment (repeatedly) [inline lib/fluent/plugin/service_discovery/round_robin_balancer.rb]: `select_node` naming is ok?
This seems out_forward's one. `select_service` is better than `select_node`?

File: lib/fluent/plugin/sd_file.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/plugin/sd_file.rb]: https://github.com/fluent/fluentd/pull/2541/commits/57d0ef02964f05cea5553ae26dbdc0e7fca9344b fixed

File: lib/fluent/plugin/sd_static.rb (hunk 3 above)
Comment (ganmacs) [reply lib/fluent/plugin/sd_static.rb]: https://github.com/fluent/fluentd/pull/2541/commits/57d0ef02964f05cea5553ae26dbdc0e7fca9344b fixed

File: lib/fluent/plugin/service_discovery/round_robin_balancer.rb (hunk 5 above)
Comment (ganmacs) [reply lib/fluent/plugin/service_discovery/round_robin_balancer.rb]: Right.  this is out_forward's name. changed to `select_service`

https://github.com/fluent/fluentd/pull/2541/commits/6a97db7dcde167fd3f1b728e2978190f8c06da0f

File: lib/fluent/plugin/sd_file.rb (hunk 2 above)
Comment (ganmacs) [reply lib/fluent/plugin/sd_file.rb]: Added information about what config needs
https://github.com/fluent/fluentd/pull/2541/commits/61a05cddf4d9918e99d71531b815237
b3f4ee905

File: lib/fluent/plugin/sd_static.rb (hunk 4 above)
Comment (ganmacs) [reply lib/fluent/plugin/sd_static.rb]: Good catch.. changed to be able to handle multiple services.
https://github.com/fluent/fluentd/pull/2541/commits/be0475893869bda7a463ed1e731a3e259ad203d5

**Comment by ganmacs** [comment]
> Should we create service_discovery helper for supporting create/configure/manage routines?

fixed https://github.com/fluent/fluentd/pull/2541/commits/9883173111ea2b2fafeee694600ab10a5cf33743

File: lib/fluent/plugin/out_forward.rb [hunk 6]
```diff
@@ -138,6 +139,10 @@ class ForwardOutput < Output
       config_param :weight, :integer, default: 60
     end
 
+    config_section :service_discovery do
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: This define should be done via helper. See `parser` helper implementation.

File: lib/fluent/plugin_helper/service_discovery.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        super
+      end
+
+      private
+
+      # @param title [Symbol] the thread name. this value should be unique.
+      # @param configurations [Hash] hash which must has discivery_service type and its configuration like `{ type: :static, conf: <Fluent::Config::Element> }`
+      # @param load_balancer [Object] object which has two methods #rebalance and #select_service
+      # @param custom_build_method [Proc]
+      def create_service_discovery_manager(title, configurations:, load_balancer: nil, custom_build_method: nil, interval: 3)
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/service_discovery.rb]: `service_disovery_create_manager` is better to follow other plugin helpers.
helpers uses its name as function prefix.

File: lib/fluent/plugin_helper/service_discovery.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        @_plugin_helper_service_discovery_title = title
+        @_plugin_helper_service_discovery_iterval = interval
+
+        @discovery_manager = Fluent::PluginHelper::ServiceDiscovery::Manager.new(
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/service_discovery.rb]: unlike other helpers, this helper doesn't allow multiple instance, right?

File: lib/fluent/plugin/service_discovery/discovery_message.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'fluent/plugin/service_discovery'
+
+module Fluent
+  module Plugin
+    class ServiceDiscovery # already loaded
+      DiscoveryMessage = Struct.new(:type, :service) do
```
Comment (repeatedly) [inline lib/fluent/plugin/service_discovery/discovery_message.rb]: Exposing this struct is needed? `ServiceDiscovery.service_in/service_out` also looks good for me.
With this approach, we can move this implementation into `service_discovery.rb` and users don't need to require additional file.

Comment (ganmacs) [reply lib/fluent/plugin/service_discovery/discovery_message.rb]: fixed https://github.com/fluent/fluentd/pull/2541/commits/4ba7c7a40e8184f6d3f685482a9f24931cd87277

File: lib/fluent/plugin/out_forward.rb (hunk 6 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: fixed https://github.com/fluent/fluentd/pull/2541/commits/54e8d537940b98c3d15c6d9d2f34e6c7790c0062

File: lib/fluent/plugin_helper/service_discovery.rb (hunk 7 above)
Comment (ganmacs) [reply lib/fluent/plugin_helper/service_discovery.rb]: fixed https://github.com/fluent/fluentd/pull/2541/commits/46e6e0c63b1e9d548d08df5e2287f511ebe43c18

File: lib/fluent/plugin_helper/service_discovery.rb (hunk 8 above)
Comment (ganmacs) [reply lib/fluent/plugin_helper/service_discovery.rb]: You mean multiple `@dicovery_mangers`? if so, yes.
This manager instance should be one per output/input plugin. the manager has multiple sd plugins.

**Comment by repeatedly** [comment]
@cosmo0920 Do you need to review this? You link this PR from elasticsearch plugin.

**Comment by cosmo0920** [comment]
I'm considering to use service discovery plugin in elasticsearch plugin.
This PR looks good to me. :+1:

> @cosmo0920 Do you need to review this? You link this PR from elasticsearch plugin.

Curently, no. The above link is just a note.

**Comment by repeatedly** [comment]
Finally merged :)

**Comment by repeatedly** [comment]
@ganmacs Could you update fluentd document for service discovery plugin?

**Pull request merged**